### DISCUSSION TITLE
feat: approval workflow for pages

### DISCUSSION THREAD

**Discussion by Lanhild** [growilabs/growi#9453] [Ideas / Requests]
A much needed feature would be an approval workflow.

As an administrator (or we could see something like an authorized group), I would like to be able to choose whether or not I approve a person's changes to a page.

GROWI using MongoDB and the principle of revisions, it could easily be leveraged to add an approval workflow to the application.

**Comment by yuki-takei** [comment]
I agree that there would be a need for such a feature in the rough concept.

The challenge is what happens if a new revision is added before the suggested revision is accepted.

1. An isolated revision with `suggestion: true` is created and proposed.
2. Under review...
3. Another user edits the same page, a new revision is created, and it becomes the latest revision.
4. The first revision with `suggestion: true` is accepted.

At this point, the conflict between revisions 3 and 4 must be resolved, which is stressful for users.

We might want to consider other ideas, such as having a mechanism like pessimistic locking, or having the suggestion as a text diff rather than a revision.

**Reply by Lanhild** [reply]
I can see it in a similar way to how pull requests work.

When a user modifies a page, it creates a revision attributed to themselves. Then, each user's revisions are listed like pull requests, and you can choose whether or not to approve one.

For revisions conflicts, I'm not sure how it can be handled.

>  such as having a mechanism like pessimistic locking, or having the suggestion as a text diff rather than a revision.

Good idea, this might also be viable.
